src/analyses/does_remove_const.cpp

```diff
@@ -0,0 +1,138 @@
+/*******************************************************************\
+
+ Module: Analyses
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <goto-programs/goto_program.h>
+#include <util/type.h>
+#include <util/expr.h>
+#include <util/std_code.h>
+#include <util/base_type.h>
+#include <ansi-c/c_qualifiers.h>
+
+#include "does_remove_const.h"
+
+/*******************************************************************\
+
+Function: does_remove_constt::does_remove_constt
+
+  Inputs:
+   goto_program - the goto program to check
+   ns - the namespace of the goto program (used for checking type equality)
+
+ Outputs:
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+does_remove_constt::does_remove_constt(
+  const goto_programt &goto_program,
+  const namespacet &ns):
+    goto_program(goto_program),
+    ns(ns)
+{}
+
+/*******************************************************************\
+
+Function: does_remove_constt::operator()
+
+  Inputs:
+
+ Outputs: Returns true if the pgroam contains a const-removing cast
+
+ Purpose: A naive analysis to look for casts that remove const-ness from
+          pointers.
+
+\*******************************************************************/
+
+bool does_remove_constt::operator()() const
+{
+  for(const goto_programt::instructiont &instruction :
+    goto_program.instructions)
+  {
+    if(instruction.is_assign())
+    {
+      const code_assignt assign=to_code_assign(instruction.code);
+      typet rhs_type=assign.rhs().type();
+      typet lhs_type=assign.lhs().type();
+
+      // Compare the types recursively for a point where the rhs is more
+      // const that the lhs
+      if(!is_type_at_least_as_const_as(lhs_type, rhs_type))
+      {
+        return true;
+      }
+
+      // see if the rhs loses const inside the expression tree
+      exprt rhs=assign.rhs();
+      while(!rhs.operands().empty())
+      {
+        typet &pre_op_type=rhs.type();
+        typet &post_op_type=rhs.op0().type();
+
+        // Types equality does not check, for example, const-ness
+        // If this is true, then this expression only modified the
+        // type by some qualifier (or not at all)
+        if(base_type_eq(pre_op_type, post_op_type, ns))
+        {
+          if(!is_type_at_least_as_const_as(pre_op_type, post_op_type))
+          {
+            return true;
+          }
+        }
+        rhs=rhs.op0();
```

Comment:
OK so I think I've done this by modifying the while loop to only descend `ID_typecast` and this seems to work (I added some more tests and they all seem fine), but I am having trouble convincing myself theoretically. For this not to work we need some C code that produces an AST where the root is not a typecast and the two sides have the same const-ness, but somewhere in the RHS the const-ness is lost. 

```
void_fp x = //(something that must also be non const)
                               \
                             probably a typecast 
                                              \ 
                                         something const
```

Maybe this isn't possible as I can't think how a typecast can appear anywhere except at the root and it's immediate descendent (or in some sub-expression like an index but that isn't related. Can you think of anything? Going to explore what happens with pointer arithmetic to see if anything can be constructed in that way.